parquet/src/main/java/org/apache/iceberg/parquet/ReadConf.java

```diff
@@ -72,29 +74,31 @@
     this.options = options;
     this.reader = newReader(file, options);
     MessageType fileSchema = reader.getFileMetaData().getSchema();
-
     boolean hasIds = ParquetSchemaUtil.hasIds(fileSchema);
-    MessageType typeWithIds = hasIds ? fileSchema : ParquetSchemaUtil.addFallbackIds(fileSchema);
 
     this.projection = hasIds ?
-        ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema) :
-        ParquetSchemaUtil.pruneColumnsFallback(fileSchema, expectedSchema);
+            ParquetSchemaUtil.pruneColumns(fileSchema, expectedSchema) :
+            ParquetSchemaUtil.pruneColumnsByName(fileSchema, expectedSchema);
```

Comment:
For Avro  we assigned ids to fileSchema based on the name mapping  and projected ids, which we derived from expected schema. [`Set<Integer> projectedIds = TypeUtil.getProjectedIds(expectedSchema)`] . This was pigged backed on column pruning code for Avro see - `AvroSchemaUtil.pruneColumns`

I'm unsure whether the same general approach can be applied here as well. @rdblue Thoughts?

Reply:
Here is a reversed way. I build name mapping based on `fileSchema`, and then project ids for expected schema according to names. The reason is the metrics are built from parquet footer so we have to bind the expression to `fileSchema` instead of expected schema in `ParquetMetricsRowGroupFilter` and `ParquetDictionaryRowGroupFilter`.